libraries/lib-concurrency/ParallelProgress.h

```diff
@@ -0,0 +1,55 @@
+/**********************************************************************
+ 
+ Audacity: A Digital Audio Editor
+ 
+ @file ParallelProgress.h
+ 
+ Paul Licameli
+ 
+ **********************************************************************/
+#ifndef __AUDACITY_PARALLEL_PROGRESS__
+#define __AUDACITY_PARALLEL_PROGRESS__
+
+#include <atomic>
+
+namespace Parallel {
+
+namespace detail { struct AbortException{}; }
+
+//! A callback supplied to run methods of a task so it can "yield" while it
+//! reports progress, also detecting total progress
+/*!
+ Throws an exception when a thread team is shutting down.  Do not handle this
+ exception.
+
+ @tparam Increment an Integral type
+ */
+template<typename Increment> class Progress {
```

Comment:
This looks like a cancellation pattern but with progress tracking. I don't think that the progress tracking is accurate or needed here.

At the same time, I think that using an exception and, specifically, an exception from the `details` namespace reduces the scope where this class is useful.